contracts/remarkables/src/contract.rs

```diff
@@ -0,0 +1,922 @@
+    );
+    Ok(Response::new()
+        .add_attribute(ATTRIBUTE_ACTION, ACTION_INSTANTIATE)
+        .add_attribute(ATTRIBUTE_SENDER, &info.sender)
+        .add_attribute(ATTRIBUTE_ADMIN, msg.admin)
+        .add_attribute(ATTRIBUTE_CW721_CODE_ID, msg.cw721_code_id)
+        .add_submessage(cw721_submessage))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut<DesmosQuery>,
+    _env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response<DesmosMsg>, ContractError> {
+    msg.validate()?;
+    match msg {
+        ExecuteMsg::MintTo {
+            post_id,
+            remarkables_uri,
+            rarity_level,
+        } => execute_mint_to(deps, info, rarity_level, post_id.into(), remarkables_uri),
+        ExecuteMsg::UpdateAdmin { new_admin } => execute_update_admin(deps, info, new_admin),
+        ExecuteMsg::UpdateRarityMintFee {
+            rarity_level,
+            new_fees,
+        } => execute_update_rarity_mint_fees(deps, info, rarity_level, new_fees),
+    }
+}
+
+fn execute_mint_to(
+    deps: DepsMut<DesmosQuery>,
+    info: MessageInfo,
+    rarity_level: u32,
+    post_id: u64,
+    remarkables_uri: String,
+) -> Result<Response<DesmosMsg>, ContractError> {
+    let rarity =
+        RARITY
+            .may_load(deps.storage, rarity_level)?
+            .ok_or(ContractError::RarityNotExists {
+                level: rarity_level,
+            })?;
+    // Check if rarity mint fees is enough
+    if !is_enough_fees(info.funds, rarity.mint_fees) {
+        return Err(ContractError::MintFeesNotEnough {});
+    }
+    // Check if post reaches the eligible threshold
+    check_eligibility(
+        deps.storage,
+        deps.querier.deref(),
+        info.sender.clone(),
+        post_id,
+        rarity.engagement_threshold,
+    )?;
+    // Create the cw721 message to send to mint the remarkables
+    let token_id = convert_post_id_to_token_id(post_id, rarity_level);
+    let mint_msg = Cw721ExecuteMsg::<Empty, Empty>::Mint(MintMsg::<Empty> {
+        token_id: token_id.clone(),
+        owner: info.sender.clone().into(),
+        token_uri: Some(remarkables_uri.clone()),
+        extension: Empty {},
+    });
+    let wasm_execute_mint_msg = wasm_execute(CW721_ADDRESS.load(deps.storage)?, &mint_msg, vec![])?;
+    Ok(Response::new()
+        .add_attribute(ATTRIBUTE_ACTION, ACTION_MINT_TO)
+        .add_attribute(ATTRIBUTE_SENDER, &info.sender)
+        .add_attribute(ATTRIBUTE_RARITY_LEVEL, rarity_level.to_string())
+        .add_attribute(ATTRIBUTE_TOKEN_ID, token_id.to_string())
+        .add_attribute(ATTRIBUTE_RECIPIENT, &info.sender)
+        .add_attribute(ATTRIBUTE_TOKEN_URI, remarkables_uri)
+        .add_message(wasm_execute_mint_msg))
+}
+
+// Returns the token id as "<post-id>-<rarity-level>"
+pub fn convert_post_id_to_token_id(post_id: u64, rarity_level: u32) -> String {
+    post_id.to_string() + "-" + &rarity_level.to_string()
```

Comment:
@bragaz I think that is not necessary, since for each instance of `remarkable` contract there's a different instance of `cw721` so such collision should never occur.
Also in the current implementation we don't allow to update the `cw721` address to one associated with another another instance of `remarkable`.

Reply:
Yes, agree with @manu0466, the collision would never occur since a `cw721` only belongs one `remarkable` contract.

Reply:
Ok cool!